contributors/design-proposals/storage/online-grow-volume-size.md

```diff
@@ -0,0 +1,361 @@
+        if err != nil {
+            return fmt.Errorf("get pod %s/%s using PVC %s failed: %v",
+                pod.Name, pod.Namespace, uniqueVolumeName, err)
+        }
+        podObj = podObj.DeepCopy()
+
+        if len(podObj.Spec.NodeName) == 0 {
+            glog.V(5).Infof("Skip MarkForFileSystemResize for pvc %s pod %s/%s: pod not scheduled",
+                uniqueVolumeName, pod.Name, pod.Namespace)
+            continue
+        }
+
+        // If volume hasn't been mounted to the node, we do not add annotation to pods running on this node.
+        if !mountedNodes[podObj.Spec.NodeName] {
+            glog.V(5).Infof("Skip MarkForFileSystemResize for pvc %s pod %s/%s: "+
+                "volume not mounted to %s yet", uniqueVolumeName, pod.Name, pod.Namespace, podObj.Spec.NodeName)
+            continue
+        }
+
+        // Check whether one or more pods has been updated in the
+        // node or not, if so, skip other pods running on same node.
+        if markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] {
+            continue
+        }
+
+        if resizeMap.podHasMarkedForPVCResizing(podObj, volumeSpec.Name()) {
+            markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] = true
+            glog.V(5).Infof("Skip update pod %s/%s annotation for "+
+                "pvc %s resizing as it already updated", podObj.Name, podObj.Namespace, uniqueVolumeName)
+        } else {
+            podObjs = append(podObjs, podObj)
+        }
+    }
+
+    for _, podObj := range podObjs {
+        if markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] {
+            glog.V(5).Infof("Skip MarkForFileSystemResize for pvc %s pod %s/%s: "+
+                "another pod in the same node has been marked", uniqueVolumeName, podObj.Name, podObj.Namespace)
+            continue
+        }
+
+        markErr := resizeMap.updatePodAnnotationForPVCResizing(podObj, volumeSpec.Name())
+        if markErr != nil {
+            glog.Errorf("Update pod %s/%s annotation for pvc %s resizing failed: %v", podObj.Name, podObj.Namespace, uniqueVolumeName, markErr)
+            return markErr
+        }
+        markedNodes[commontypes.NodeName(podObj.Spec.NodeName)] = true
+    }
+
+    return nil
+}
+```
+
+### File System Resize on Kubelet
+
+#### Prerequisite of Online File System Resize
+
+* Pod running on the node has one more `volumeFSResizingAnnotation` annotations.
+* Volume with a online resizing request has been mounted to the node.
+* `PV.Spec.Capacity` is greater than `PVC.Status.Capacity`.
+
+#### Volume Resize Request Detect
+
+Kubelet need to periodically loop through the list of active pods to detect which volume requires file system resizing. We use a `VolumeToResize` object to represent a volume resizing requests and introduce a new interface `VolumeFsResizeMap` rather than `DesiredStateOfWorld` to cache such `VolumeToResize` objects, since we treat it as a request queue, not a state. This interface will look like this:
```

Comment:
Kubelet already loops through all mounts in VolumeManager. It can see that a volume needs fs resize there. And based on whether the mount exists in its ASW, it can see if it's going to perform online or offline resize.

In addition, fs resize must be coordinated with VolumeManager anyway, since VolumeManager should not touch the volume (e.g. unmount it) during resize.

Reply:
> Kubelet already loops through all mounts in VolumeManager. It can see that a volume needs fs resize there.

The problem is that the volume objects cached in `ASW` is only updated when pod starting, and won't update after pod started any more, it means that if we resize the volume after pod already started, `VolumeManager` can't realize this change from `ASW`.

> In addition, fs resize must be coordinated with VolumeManager anyway, since VolumeManager should not touch the volume (e.g. unmount it) during resize.

We append the resize operation to `OperationExecutor` so that `VolumeManager` can't touch the volume before resize operation finished.

Reply:
The original discussion in this issue: kubernetes/kubernetes#57185

Reply:
I think - what we can do is. When we are iterating through pods in volumemanager's desired_state_of_world for mount and call `PodExistsInVolume` function, then that function can make an additional API call to check if underlying volume needs resizing.  Currently `PodExistsInVolume` already raises an error `remountRequired` when some change requires us to remount the volume.  So we can introduce a new error `resizeRequired` which will be raised by the function and that in turn will cause file system resize on the kubelet. 

This will require no API change or annotation and everything will work online as expected. thanks @jsafrane for pointing it out.

Reply:
One potential problem with ^ approach is - the reconciler loop turns every 100ms and say if a node has 20 pods with volumes, then we will be making 20 GET requests every 100ms, because there is no informer inside volume_manager. Again - I am not sure how badly it will affect us.

Reply:
Maybe we can add a `lastUpdateTimestamp` for each volumes, and use an independent loop interval with the `reconciler`, this can reduce the request counts, but I am not sure how much help this can have